File: src/py/agent_studio/providers/diagnostics.py
"""Format health reports and errors for the CLI (secret-safe)."""

from __future__ import annotations

from ..core.redact import scrub_string
from .errors import ProviderError
from .testing import HealthReport

_ICON = {"ok": "\u2713", "warn": "!", "error": "\u2717"}

_REMEDIATION = {
    "missing_api_key": "Add the key to .env.local (git-ignored), then re-run the check.",
    "invalid_api_key": "Double-check the API key value and that it has access to this endpoint.",
    "not_found": "Verify the base URL path (it usually ends in /v1) and the model id.",
    "rate_limit": "You are being rate limited. Wait a moment and try again.",
    "timeout": "The endpoint took too long. Check connectivity or increase the timeout.",
    "network": "Check the base URL, your network/VPN, and any firewall/proxy.",
    "server": "The provider had a server-side error. Retry shortly.",
    "invalid_config": "Review the provider configuration (base URL, kind, model).",
}


def format_health_report(report: HealthReport) -> str:
    lines = [
        f"Provider:  {report.label} ({report.provider_id}) [{report.kind}]",
        f"Endpoint:  {report.base_url or '(not set)'}",
        f"Model:     {report.model or '(not set)'}",
        f"API key:   {report.api_key_masked}"
        + (f" (from {report.api_key_env})" if report.api_key_env else ""),
        f"Overall:   {report.overall.upper()}",
        "",
        "Checks:",
    ]
    for c in report.checks:
        lines.append(f"  {_ICON.get(c.status, '?')} {c.name}: {c.message}")
    if report.latency_ms is not None:
        lines.append("")
        lines.append(f"Latency:   {int(report.latency_ms)}ms")
    return scrub_string("\n".join(lines))


def format_error(err: ProviderError) -> str:
    remedy = _REMEDIATION.get(err.kind, "Review the configuration and try again.")
    return scrub_string(f"{err.message}\nWhat to do: {remedy}")

File: src/engine/providers/diagnostics.ts
// Formats health reports and provider errors into user-friendly, secret-safe
// diagnostics for the CLI and web clients.

import { scrubString } from '../core/redact.ts';
import type { ProviderError, ProviderErrorKind } from './errors.ts';
import type { CheckStatus } from './types.ts';
import type { HealthReport } from './testing.ts';

const STATUS_ICON: Record<CheckStatus, string> = { ok: '✓', warn: '!', error: '✗' };

/** Actionable remediation hints keyed by error kind. */
export function remediationFor(kind: ProviderErrorKind, envName?: string): string {
  switch (kind) {
    case 'missing_api_key':
      return envName
        ? `Add ${envName}=<your-key> to .env.local (it is git-ignored), then re-run the test.`
        : 'Provide an API key for this provider, then re-run the test.';
    case 'invalid_api_key':
      return 'Double-check the API key value and that it has access to this endpoint.';
    case 'not_found':
      return 'Verify the base URL path (it usually ends in /v1) and that the model id is correct.';
    case 'rate_limit':
      return 'You are being rate limited. Wait a moment and try again, or reduce request frequency.';
    case 'timeout':
      return 'The endpoint took too long. Check connectivity, increase the timeout, or try again.';
    case 'network':
      return 'Check the base URL, your internet/VPN connection, and any firewall/proxy settings.';
    case 'server':
      return 'The provider had a server-side error. This is usually temporary — retry shortly.';
    case 'invalid_config':
      return 'Review the provider configuration (base URL, kind, model).';
    default:
      return 'Review the configuration and try again.';
  }
}

/** Render a health report as plain text lines (already secret-safe). */
export function formatHealthReport(report: HealthReport): string {
  const lines: string[] = [];
  const p = report.provider;
  lines.push(`Provider:  ${p.label} (${p.id}) [${p.kind}]`);
  lines.push(`Endpoint:  ${p.baseUrl || '(not set)'}`);
  lines.push(`Model:     ${p.model || '(not set)'}`);
  lines.push(`API key:   ${p.apiKeyMasked}${p.apiKeyEnv ? ` (from ${p.apiKeyEnv})` : ''}`);
  lines.push(`Overall:   ${report.overall.toUpperCase()}`);
  lines.push('');
  lines.push('Checks:');
  for (const c of report.checks) {
    lines.push(`  ${STATUS_ICON[c.status]} ${c.name}: ${c.message}`);
  }
  if (report.latencyMs != null) {
    lines.push('');
    lines.push(`Latency:   ${report.latencyMs}ms`);
  }
  return scrubString(lines.join('\n'));
}

/** One-line summary of a provider error plus remediation. */
export function formatError(err: ProviderError, envName?: string): string {
  return scrubString(`${err.message}\nWhat to do: ${remediationFor(err.kind, envName)}`);
}
